patches/tModLoader/Terraria/ModLoader/ExtraJumpLoader.cs

```diff
@@ -0,0 +1,224 @@
+
+		// Sort the modded jumps per slot
+		List<ExtraJump> sorted = new();
+
+		for (int i = 0; i < DefaultExtraJumpCount + 2; i++) {
+			var sort = new TopoSort<ExtraJump>(sortingSlots[i].Select(static t => ExtraJumps[t]),
+				j => positions[j.Type].OfType<ExtraJump.After>().Select(static a => a.Parent).OfType<ExtraJump>(),
+				j => positions[j.Type].OfType<ExtraJump.Before>().Select(static b => b.Parent).OfType<ExtraJump>());
+
+			foreach (ExtraJump jump in sort.Sort()) {
+				sorted.Add(jump);
+			}
+
+			if (i < DefaultExtraJumpCount)
+				sorted.Add(ExtraJumps[i]);
+		}
+
+		orderedJumps = sorted.ToArray();
+	}
+
+	private static ExtraJump.Position VerifyPositionType(ExtraJump jump, ExtraJump.Position position) {
+		if (position is not ExtraJump.After and not ExtraJump.Before)
+			throw new ArgumentException($"ModExtraJump {jump} has unknown Position {position}");
+
+		return position;
+	}
+
+	internal static void RegisterDefaultJumps()
+	{
+		int i = 0;
+		foreach (var jump in ExtraJumps) {
+			jump.Type = i++;
+			ContentInstance.Register(jump);
+			ModTypeLookup<ExtraJump>.Register(jump);
+		}
+	}
+
+	public static void ModifyPlayerHorizontalSpeeds(Player player)
+	{
+		foreach (ExtraJump moddedExtraJump in orderedJumps) {
+			// Special case: Sandstorm in a Bottle uses a separate flag
+			ref ExtraJumpData extraJump = ref player.GetExtraJump(moddedExtraJump);
+			if ((object.ReferenceEquals(moddedExtraJump, ExtraJump.SandstormInABottle) && player.sandStorm) || (extraJump.PerformingJump && extraJump.Active))
+				moddedExtraJump.ModifyHorizontalSpeeds(player);
+		}
+	}
+
+	public static void JumpVisuals(Player player)
+	{
+		foreach (ExtraJump jump in orderedJumps) {
+			ref ExtraJumpData data = ref player.GetExtraJump(jump);
+			if (data.PerformingJump && data.Active && !data.JumpAvailable && jump.PreVisuals(player) && PlayerLoader.PreExtraJumpVisuals(jump, player)) {
+				jump.Visuals(player);
+				PlayerLoader.ExtraJumpVisuals(jump, player);
+			}
+		}
+	}
+
+	public static void ProcessJumps(Player player, bool flipperOrSlimeMountSwimming)
+	{
+		if (!flipperOrSlimeMountSwimming) {
+			foreach (ExtraJump jump in orderedJumps) {
+				ref ExtraJumpData data = ref player.GetExtraJump(jump);
+				if (data.JumpAvailable) {
+					data.JumpAvailable = false;
+					data.PerformingJump = true;
+					jump.PerformJump(player);
+					break;
+				}
+			}
+
+			// The Basilisk mount's extra jump is always "consumed", even if a higher-priority jump was performed
+			player.GetExtraJump(ExtraJump.BasiliskMount).JumpAvailable = false;
+		}
+	}
+
+	public static void OnJumpEnded(Player player)
+	{
+		foreach (ExtraJump jump in orderedJumps) {
+			ref ExtraJumpData data = ref player.GetExtraJump(jump);
+			if (data.PerformingJump) {
+				jump.OnEnded(player);
+				PlayerLoader.OnExtraJumpEnded(jump, player);
+				data.PerformingJump = false;
+			}
+		}
+	}
+
+	public static void RefreshExtraJumps(Player player)
```

Comment:
vanilla has one case where they can be publicly refreshed (`Player::RefreshMovementAbilities`).  every other case is either from standing on the ground or jumping out of a grapple.
should that Player method be made `internal` along with this ExtraJumpLoader method?

Reply:
The `Active` check here means that if something is suppressing a double-jump when the player leaves the ground, but then expires, the jump will not be considered available :(

Reply:
alas, this is a mimicry of vanilla behavior

granted, vanilla never had anything except for a hardcoded flipper check for disabling jumps.  this could probably be remedied by the `Disable()` method stuff mentioned in another conversation